Adjust comments to sparse covariance computation and fix the complex case

andreasnoack opened the pull request.

andreasnoack committed:
Adjust comments to sparse covariance computation. Use sample
instead of population covariance in comments since that is what
the function computes.

Sacha0 reviewed (approved):
superficially lgtm! :)

tkelman commented on base/sparse/linalg.jl:
```diff
@@ -946,17 +946,17 @@ function Base.cov(X::SparseMatrixCSC, vardim::Int=1; corrected::Bool=true)
+    # Compute (estimate of) nμ
     sums = sum(X, vardim)
+
+    # Subtract nμμ' from X'X
```
does unscaled_covzm compute `X'X` or `nX'X` ?

andreasnoack replied:
It computes `X'X` or an estimate of `nE(XX')`. The notation is a little confusing since `X` is a vector in the population expression above: `E(XX')` whereas it here refers to the data matrix where each row is a vector similar to the original `X` vector. Hence `X'X` (where `X` is matrix) is an estimate of `nE(XX')` (where `X` is vector). The quantity `μ=EX` is the actual population mean (which is unknown) so we subtract an *estimate* of `nμμ'`. I included the *estimate* part in the comment above but decided it wasn't necessary to repeat here. I think people with a little stats background will understand the comment in the current form. Making it precise will either require explicit sum notation with subscripts which is verbose or require that we stop using the familiar population notation `Cov(X)=E(X-μ)(X - μ)'` and instead write the sample version in matrix form `1/(n-1)(X .- x̄)'(X .- x̄)` with `x̄ = mean(X, 1)` which is not how people typically express the covariance.

tkelman replied:
I have enough stats background to have seen (and taught...) covariance of matrices before and the notation collision is still confusing, especially when trying to describe code

Sacha0 replied:
Some additional explanation like that in your (`@andreasnoack`'s) response above would be a lovely clarification: Like Tony, evaluating this required meaningful deciphering on my part. Best!

andreasnoack replied:
I think this can be made a bit simpler. I'll update the PR later today.

andreasnoack committed:
Fix the sparse covariance with complex elements and simplify tests

andreasnoack commented:
I realized that the complex case wasn't handled correctly so I've fixed that as well. It was just a single conjugation. When adding tests for this, I made the tests loop over `vardim` and `corrected` instead of unrolling and I changed the check of NaN and Inf propagation to testing the `isfinite` pattern.

tkelman commented on base/sparse/linalg.jl:
```diff
@@ -946,17 +946,21 @@ function Base.cov(X::SparseMatrixCSC, vardim::Int=1; corrected::Bool=true)
+    # Compute x̄
+    x̄ᵀ = mean(X, vardim)
+
+    # Subtract n*x̄*x̄' from X'X
```
the transpose in the mean variable name isn't exactly consistent with this

andreasnoack replied:
Why not? `x̄ᵀ[i] == x̄[i]`.

nalimilan commented on test/sparse/sparse.jl:
```diff
@@ -1853,120 +1853,49 @@ end
+    @testset "Test with no Infs and NaNs, vardim=$vardim, corrected=$corrected" for vardim in (1, 2),
+                                                                                 corrected in (true, false)
+        @test cov(x_sparse, vardim, corrected=corrected) ≈
+                cov(x_dense, vardim, corrected=corrected)
```
Incorrect alignment here and below. Same for `in` below.

andreasnoack replied:
It is 16. Do you want 12 or aligned with the `cov` above?

nalimilan replied:
I would say either 4 spaces from the previous level, or aligned with the `cov` above. Not sure which convention is more common.

Sacha0 commented on base/sparse/linalg.jl:
```diff
@@ -946,17 +946,21 @@ function Base.cov(X::SparseMatrixCSC, vardim::Int=1; corrected::Bool=true)
 
-    # Cov(X) = E[(X-μ)'(X-μ)]
-    # = X'X - X'μ
+    # The ovariance can be decomposed into two terms
```
I am not familiar with this `ovariance` concept :).

Sacha0 commented on base/sparse/linalg.jl:
```diff
@@ -946,17 +946,21 @@ function Base.cov(X::SparseMatrixCSC, vardim::Int=1; corrected::Bool=true)
-    # = X'X - X'μ
+    # The ovariance can be decomposed into two terms
+    # 1/(n - 1) ∑ (x_i - x̄)*(x_i - x̄)' = 1/(n - 1) (∑ x_i*x_i' - n*x̄*x̄')
+    # which allows for sparse matrix matrix product
```
Perhaps "which can be evaluated via a sparse matrix-matrix product"?
